src/platforms/common/stm32/serialno.c

```diff
@@ -43,7 +45,13 @@ void read_serial_number(void)
 #elif defined(STM32L0) || defined(STM32F0) || defined(STM32F3)
 	int offset = 5;
 #endif
-	sprintf(serial_no, "%04X%04X%04X", uid[1] + uid[5], uid[0] + uid[4], uid[offset]);
+	static char buf[8];
+	utoa_upper(uid[1] + uid[5], buf, 16);
```

Comment:
Given how this function works internally, perhaps move the static character buffer into the function, and have it take the serial number buffer at the appropriate offset for the next chunk - that eliminates the `strncpy()` (which itself is a lint item for complex reasons to do with its operation - in this use-case, we'd ask for `memcpy()` instead) and allows this to become something like:
```c
utoa_upper(uid[1] + uid[5], serial_no + 0, 16);
utoa_upper(uid[0] + uid[1], serial_no + 4, 16);
utoa_upper(uid[offset], serial_no + 8, 16);
```

Note that serial_no is already a pointer for the purposes of all this code due to array -> pointer decay, so the `&serial_no[0]` syntax is just convoluted - `serial_no + 0` does exactly the same thing but more obviously.

Reply:
Do you mean the implementation can initially convert into the fixed static buffer (with no malloc, which is good for embedded) and then reverse into callsite-provided nonnul *buf? I like this idea, even if we'd need to put some range bounds guardrail checks (for e.g. asking a base-2 of a uint32_t worst case the buffer has to be char[33])

Reply:
Yes, that's exactly what we mean 🙂 think it'll also improve the codegen of the reversal step too.

Reply:
Pinned a buffer to the function, which is okay for our purposes (but state is a no-no in libc functions, see Re-entrancy). Could be a VLA, but it's a bootloader, so nope.
Updated callsite per suggestion, eliminating both strncpy and outside buffer, tested live.

Reply:
To be fair, the buffer could be non-static inside the function which eliminates the statefullness problem - at 33 bytes of stack (32 if we eliminate the extra nul termination reservation), it's not a huge deal.

Reply:
Sorry, my bad at saying char[33] is large. At bootloader scope -- suggestion declined. This firmware currently occupies ... (checks Puncover)
```
prog 	1,036
usbd_control_buffer 	1,024
usbd_dev 	276
if_string 	64
buf 	33
serial_no 	13
magic 	8
...
   text    data     bss     dec     hex filename
   8512      84    2404   11000    2af8 src/blackmagic_dfu
```
not much, and using measly 33 (rounded up for any alignment) bytes is insignificant in comparison to dangers of stack allocation. Even for a hypothetical F1 target ram usage of 20 KiB would be ~12%, and this actually targets F4 chips which have more. I would not want people to have doubts that this software component is stable or not. Is this critical?

Reply:
The problem we foresee is that the component is shared with the firmware were it's eating into the heap availability for something that's used once during startup (during setup of the USB descriptors) when, in both cases, it's called with a shallow call stack and something like 4KiB of stack availability. It's a fair comment to wonder if this might cause stack usage issues though 🤔

Reply:
Given how the code now looks, the main ask here is to make `buf` 32 chars long as you've eliminated the need for the terminator - otherwise we'll take it as-is

Reply:
Resized to 32, works without changes in an external testcase and live on device (we don't ask to convert to binary anyway... yet)